### ISSUE TITLE
SDPA for gpt_bigcode/starcoder

### ISSUE THREAD

**Issue by sentialx** [huggingface/optimum#1249]
### Feature request

Please add fast path attention support for gpt_bigcode family models (santacoder, starcoder).

### Motivation

Model being slow without SDPA

### Your contribution

I tried to implement it by myself, but I'm not even sure if PyTorch's SDPA supports MQA.

**Comment by fxmarty** [comment]
Than you - will do today

**Comment by fxmarty** [comment]
Hi @sentialx, flash attention does not support fp32, so the `attention_softmax_in_fp32` from starcoder will not be usable (see e.g. https://huggingface.co/bigcode/starcoder/blob/main/config.json#L7). I'll still integrate with a warning, but if you are using fp16, do not expects logits to match due to this.

**Comment by fxmarty** [comment]
Hi @sentialx, the implementation was merged. I did not benchmark it for speed though, feel free to try one of the tests in https://github.com/huggingface/optimum/tree/main/tests/benchmark

**Comment by sentialx** [comment]
Thank you @fxmarty! It is much faster now.

**Comment by anmolagarwal999** [comment]
@sentialx What is the factor of speed gain you are noticing ?

**Comment by fxmarty** [comment]
@sentialx @anmolagarwal999 you'll need this PR if you want to dispatch on flash/mem-efficient attention: https://github.com/huggingface/optimum/pull/1259

I am getting ~35% speedup for inference on an A100, did not try training.

**Comment by anmolagarwal999** [comment]
> @sentialx @anmolagarwal999 you'll need this PR if you want to dispatch on flash/mem-efficient attention: #1259
> 
> I am getting ~35% speedup for inference on an A100, did not try training.

When can we expect this to be merged with the main branch ?

**Comment by fxmarty** [comment]
It is now!
